certbot-nginx/certbot_nginx/parser_obj.py

```diff
@@ -0,0 +1,407 @@
+
+    :returns: whether parsed_obj represents a comment sentence.
+    :rtype bool:
+    """
+    if not isinstance(parsed_obj, Sentence):
+        return False
+    return parsed_obj.words[0] == "#"
+
+def _is_certbot_comment(parsed_obj):
+    """ Checks whether parsed_obj is a "managed by Certbot" comment.
+
+    :param .WithLists parsed_obj:
+
+    :returns: whether parsed_obj is a "managed by Certbot" comment.
+    :rtype bool:
+    """
+    if not _is_comment(parsed_obj):
+        return False
+    if len(parsed_obj.words) != len(COMMENT_BLOCK):
+        return False
+    for i, word in enumerate(parsed_obj.words):
+        if word != COMMENT_BLOCK[i]:
+            return False
+    return True
+
+def _certbot_comment(context, preceding_spaces=4):
+    """ A "Managed by Certbot" comment.
+    :param int preceding_spaces: Number of spaces between the end of the previous
+        statement and the comment.
+    :returns: Sentence containing the comment.
+    :rtype: .Sentence
+    """
+    result = Sentence(context)
+    result.parse([" " * preceding_spaces] + COMMENT_BLOCK)
+    return result
+
+def _choose_parser(child_context, list_):
+    """ Choose a parser from child_context, based on whichever hook returns True first. """
+    for type_ in child_context.parsing_hooks:
+        if type_.should_parse(list_):
+            return type_(child_context)
+    raise errors.MisconfigurationError(
+        "None of the parsing hooks succeeded, so we don't know how to parse this set of lists.")
+
+def parse_raw(lists_, context=None, add_spaces=False):
+    """ Primary parsing factory function. Based on `context.parsing_hooks`, chooses
+    WithLists objects with which it recursively parses `lists_`.
+
+    :param list lists_: raw lists from pyparsing to parse.
+    :param .ParseContext context: Context containing parsing hooks. If not set,
+        uses default parsing hooks.
+    :param bool add_spaces: Whether to pass add_spaces to the parser.
+
+    :returns .WithLists: The parsed object.
+
+    :raises errors.MisconfigurationError: If no parsing hook passes, and we can't
+        determine which type to parse the raw lists into.
+    """
+    if context is None:
+        context = ParseContext()
+    if context.parsing_hooks is None:
+        context.parsing_hooks = DEFAULT_PARSING_HOOKS
+    parser = _choose_parser(context, lists_)
+    parser.parse(lists_, add_spaces)
+    return parser
+
+# Default set of parsing hooks. By default, lists go to Statements.
+DEFAULT_PARSING_HOOKS = (Block, Sentence, Statements)
+
+class ParseContext(object):
+    """ Context information held by parsed objects.
+
+    :param .WithLists parent: The parent object containing the associated object.
+    :param tuple parsing_hooks: Parsing order for `parse_raw` to use.
+    """
+    def __init__(self, parent=None, parsing_hooks=DEFAULT_PARSING_HOOKS):
+        self.parsing_hooks = parsing_hooks
+        self.parent = parent
+
+    def child(self, parent):
```

Comment:
`create_child`?

Reply:
actually, why is `parent` on the context, rather than in the object? is `ParseContext` just the metadata for a `WithLists`? If so, let's put it inside the class, and if not, what's going on here?

Reply:
oh, it is so we can just pass around the contexts for the helper methods?

Reply:
Great question! I introduced contexts in the original giant PR to keep track of things like filesystem location (so we could write these objects to the appropriate files afterwards), but am now questioning whether we need them at all-- At the very least, we definitely don't need them in this PR, so I ripped them out. If it turns out something like this would be helpful for the follow-up PRs, I will introduce them in those PRs :)